[NFC] Allow errors to be conditionally emitted as warnings

beccadax opened the pull request:
Sometimes, the same diagnostic should be emitted as an error in some situations and a warning in others; for instance, this is often done for backwards compatibility in older language modes. Previously, the warning and error had to be two separate diagnostics with their own IDs and their own entries in the diagnostic tables.

This PR adds an `InFlightDiagnostic::limitBehavior()` method which can be used to "demote" an error to a warning. More broadly, it can actually be used to transform any kind of diagnostic to a less "severe" behavior:

Fatal error → Error → Warning → Remark → Note → Ignored (no diagnostic)

This will allow us to avoid adding and maintaining redundant warnings. To test it, I have converted some recently-added code in TypeCheckConcurrency.cpp to use the feature.

As part of this work, I have also removed `DiagnosticEngine::ignoreDiagnostic()`. This feature has apparently never been used, and it was implemented using a very general mechanism that would require some work to reconcile with `limitBehavior()`. Since there are currently no examples of how `ignoreDiagnostic()` is supposed to be used, I chose to remove it rather than try to work on code I cannot test.

beccadax committed:
[NFC] Un-nest DiagnosticState::Behavior

beccadax committed:
[NFC] Pass whole diagnostic into determineBehavior()

beccadax commented:
@swift-ci please smoke test

beccadax committed:
[NFC] Rework DiagnosticEngine::ignoreDiagnostic()

This feature was implemented with a per-diagnostic table of override DiagnosticBehaviors. That table was able to represent not only that a diagnostic should be ignored, but also many other states that were difficult to integrate with a feature I’m trying to introduce. Simplify the implementation so it can only represent ignoring a diagnostic, which has a much cleaner interaction with the new feature.

beccadax commented:
LLDB was using `ignoreDiagnostic()`, so I've reimplemented it in a way that interacts more cleanly with `limitBehavior()`.

@swift-ci please smoke test

beccadax commented:
@swift-ci please smoke test Linux platform

beccadax commented:
@swift-ci please smoke test Linux platform

hborla commented on lib/AST/DiagnosticEngine.cpp:
```diff
@@ -850,27 +857,34 @@ DiagnosticBehavior DiagnosticState::determineBehavior(const Diagnostic &diag) {
   //      that
-  if (diagInfo.kind == DiagnosticKind::Warning) {
+  if (diagInfo.kind == DiagnosticKind::Warning
+      || (diag.getBehaviorLimit() == DiagnosticBehavior::Warning && !isNote)) {
```
Is this going to end up suppressing/erroring if the diagnostic kind is a `Remark` and the behavior limit is `Warning`?

beccadax replied:
You're right, this is wrong.

beccadax replied:
The whole method was kind of confusing, so I've rewritten it so that we first determine the behavior that's being requested and then apply all of these state- and settings-related changes. It also now has more straightforward control flow, so hopefully this version will be better.

hborla replied:
Yes, it's definitely much easier to read now!

beccadax committed:
[NFC] Reorganize DiagnosticState::determineBehavior()

This makes it a little easier to reason abut what’s going on while also fixing a small, currently-unexercisable flaw that would have previously occurred if you used limitBehavior(DiagnosticBehavior::Warning) and emitted a remark.

beccadax commented:
@swift-ci please smoke test

beccadax committed:
[NFC] Add InFlightDiagnostic::limitBehavior()

This allows code to reduce the “severity” of a diagnostic at one particular emission site, so that (for instance) an error can be emitted as a warning in some situations. It also changes some code in TypeCheckConcurrency.cpp to use this new feature, allowing us to delete some redundant diagnostic definitions.

beccadax commented:
Rebased due to changes in TypeCheckConcurrency.cpp.

beccadax commented:
@swift-ci please smoke test and merge

beccadax commented:
@swift-ci please smoke test macOS platform and merge

beccadax commented:
@swift-ci please smoke test macOS platform
